### PR TITLE
Fix registration of GraphQL application classes for reflection

### LINKED ISSUE #11452: SmallRye GraphQL reflection warnings
I got these warnings when building the deployment module of the SmallRye GraphQL extension:

```
[INFO] Running io.quarkus.smallrye.graphql.deployment.GraphQLTest
[io.quarkus.smallrye.graphql.deployment.TestResource, double, io.quarkus.smallrye.graphql.deployment.TestRandom, java.lang.String, io.quarkus.smallrye.graphql.deployment.TestPojo]
2020-08-18 19:03:04,049 WARN  [io.qua.dep.QuarkusAugmentor] (main) Using Java versions older than 11 to build Quarkus applications is deprecated and will be disallowed in a future release!
2020-08-18 19:03:04,969 WARN  [io.qua.dep.ste.ReflectiveHierarchyStep] (build-24) Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
	- io.opentracing.Span (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.opentracing.Tracer (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.smallrye.graphql.api.Context (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.smallrye.graphql.schema.model.Argument (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.smallrye.graphql.schema.model.Field (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.smallrye.graphql.schema.model.InterfaceType (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- io.smallrye.graphql.schema.model.Operation (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.dataloader.DataLoaderRegistry (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.ConcurrentGauge (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Counter (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Gauge (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Histogram (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Metadata (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Meter (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Metered (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Metric (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.MetricID (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.MetricRegistry (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.MetricType (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.SimpleTimer (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.SimpleTimer$Context (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Snapshot (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Timer (source: io.quarkus.smallrye.graphql.deployment.TestResource)
	- org.eclipse.microprofile.metrics.Timer$Context (source: io.quarkus.smallrye.graphql.deployment.TestResource)
```

The issue is that we create a `ReflectiveHierarchyBuildItem` for the whole `TestResource` and it contains this method:
```
    public GraphQLSchema.Builder addMyOwnEnum(@Observes GraphQLSchema.Builder builder) {

        GraphQLEnumType myOwnEnum = GraphQLEnumType.newEnum()
                .name("SomeEnum")
                .description("Adding some enum type")
                .value("value1")
                .value("value2").build();

        return builder.additionalType(myOwnEnum);
    }
```

and I think it's registering `GraphQLSchema.Builder` for reflection that brings all these classes (but it would need to be verified, there might be another additional culprit)

### PR DESCRIPTION
**Pull request by phillip-kruger**
Fix #11452

Signed-off-by:Phillip Kruger <phillip.kruger@gmail.com>

### CODE REVIEW

**Comment by gsmet** [comment]
Please don't add the 1.7 version, I do it when backporting.

**Comment by gsmet** [comment]
Are the return types of the GraphQL methods still properly registered for reflection? I would suppose that's what's needed?

**Comment by phillip-kruger** [comment]
Yes, return types and arguments, and they are, that is what that method is doing, that line that I removed should not have been there, as that is the owner class that is not needed.

**Comment by gsmet** [comment]
OK, perfect!

**Comment by phillip-kruger** [comment]
@gsmet - there is an issue with the IT. Looking at it now...

**Comment by phillip-kruger** [comment]
Ok, initial attempt was the wrong fix. We still need to register the owner class, as the framework use reflection on the methods.
I changed the way the classes are register and that seems to do the trick.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java [hunk 1]
```diff
@@ -156,11 +155,7 @@ void buildExecutionService(
         recorder.createExecutionService(beanContainer.getValue(), schema);
 
         // Make sure the complex object from the application can work in native mode
-        for (String c : getClassesToRegisterForReflection(schema)) {
-            DotName name = DotName.createSimple(c);
-            org.jboss.jandex.Type type = org.jboss.jandex.Type.create(name, org.jboss.jandex.Type.Kind.CLASS);
-            reflectiveHierarchyProducer.produce(new ReflectiveHierarchyBuildItem(type, index));
-        }
+        reflectiveClassProducer.produce(new ReflectiveClassBuildItem(true, true, getSchemaJavaClasses(schema)));
```
Comment (gsmet) [inline extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:158]: I'm not sure it's the right fix.

What I think would be right is:
- register the owner with a class build item
- register the return types of the GraphQL methods with a hierarchy build item

If you have a hierarchy of complex objects for the return type, I think what you did will not work.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a question/suggestion. Let's sort it out before merging.

File: extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java (hunk 1 above)
Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:158]: We already know the hierarchy as the model build that, but I am happy to change as you suggest. Let me add a test case with that scenario then we can see.

Comment (phillip-kruger) [reply extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java:158]: Hey @gsmet - I added some tests to test parent and child classes. All seems ok. The model we build already discover all the classes needed. So this fix should be ok. Let me know.

**Review by gsmet — APPROVED** [review]
OK, let's try that then.

**Comment by phillip-kruger** [comment]
Thanks @gsmet !

**Pull request merged**